### DISCUSSION TITLE
Allow plugins to be customizable

### DISCUSSION THREAD

**Discussion by timkley** [alpinejs/alpine#3275] [3. Feature ideas]
Hi community and @calebporzio 👋🏻 

today I've noticed that it currently seems to be impossible to configure plugins in user land.

What about accepting a second parameter when registering a plugin, like this:

```
Alpine.plugin(persist, { defaultStorage: localStorageWithTTL })
```

**Edit:** As @ekwoka pointed out, the following syntax is probably the better one, as it allows plugin developers to provide better context for type safety and would also make the impact smaller, as only plugins that need configuration need to be changed:

```
Alpine.plugin(persist({ defaultStorage: localStorageWithTTL })
```

This also outlines my use case I tried solving today: we use the `persist` plugin mainly to persist data which is bound to your user, not necessarily to the browser session. The user has it's own TTL which is provided by the server.

So, instead of updating every reference to `$persist` I'd like to overwrite default session storage.

If this is something that's likely to be accepted I'd invest time into a PR :)

**Comment by ekwoka** [comment]
Probably smart.

The more appropriate implementation would be `Alpine.plugin(persist(config))` probably.

**Reply by timkley** [reply]
I don't see any pros or cons for one way or the other. Since other projects like [Vue](https://vuejs.org/guide/reusability/plugins.html#writing-a-plugin) or [11ty](https://www.11ty.dev/docs/plugins/#plugin-configuration-options) use the other syntax it's maybe better DX to use that?

**Reply by ekwoka** [reply]
The issue with the first is that it takes a larger adjustment to how plugin works and would probably struggle to find a way to make it type safe.

The composable higher order function method allows plugins to all opt in to configurations with easy type safety.

**Reply by timkley** [reply]
Thank you for the explanation, I hadn't thought about it this way. I'll edit my original post to take this into account :)

**Reply by ekwoka** [reply]
Probably should have remembered this: https://www.npmjs.com/package/@ekwoka/x-rias

That's a plugin I had made that uses this setup
